packages/iov-bns/src/types.ts

```diff
@@ -214,6 +214,18 @@ export interface AddAddressToUsernameTx extends UnsignedTransaction {
   readonly payload: ChainAddressPair;
 }
 
+export interface Participant {
+  readonly signature: Uint8Array | null;
+  readonly power: number | null;
```

Comment:
What's the meaning of `signature: null` or `power: null`? I think this should be documented.

Maybe these cases never exist and the nullability is only for the protobuf type. In this case we should assert them away and expose non-nullable properties in IOV-Core.

Reply:
Based on this I think we can make them non-nullable: https://github.com/iov-one/weave/blob/cb5386a6e341eeb26237754a9928000825a72065/x/multisig/codec.proto#L23-L26

Reply:
They are required in the validation logic. Protobuf generation allows null and undefined, which map to missing fields in the binary representation. For higher-level types we can declare validity with the types.

Here are relevant lines that enforce it:

There must be between 1 and 100 participants: https://github.com/iov-one/weave/blob/cb5386a6e341eeb26237754a9928000825a72065/x/multisig/msg.go#L25-L33

Signature and Power must both be valid: https://github.com/iov-one/weave/blob/cb5386a6e341eeb26237754a9928000825a72065/x/multisig/msg.go#L66-L73 (and thresholds must be reasonable numbers)

Power must be between 1 and 255: https://github.com/iov-one/weave/blob/cb5386a6e341eeb26237754a9928000825a72065/x/multisig/model.go#L18-L34 (by the way this is renamed to weight in current master to be compatible with the type)

Signature (Address) must be 20 bytes: https://github.com/iov-one/weave/blob/cb5386a6e341eeb26237754a9928000825a72065/conditions.go#L206-L214

Fit as much as you can in the type system. The rest gladly in run-time checks (or just let the blockchain feed you errors)

Reply:
Thanks for the feedback